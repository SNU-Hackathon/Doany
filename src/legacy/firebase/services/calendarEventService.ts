/**
 * Calendar Event Service
 * 
 * 경계/정합성 규칙:
 * 1. 기간이 7일 미만이면 검증 스킵 (불완전 주만 존재)
 * 2. 중복 이벤트 카운트 정책: 동일 날짜·시간 다중 등록은 개수만큼 집계 (기본)
 * 3. 타임존: Asia/Seoul 고정, 날짜 문자열은 YYYY-MM-DD
 */

import { collection, doc, getDocs, orderBy, query, serverTimestamp, where, writeBatch } from 'firebase/firestore';
import { v4 as uuidv4 } from 'uuid';
import { CalendarEvent } from '../types';
import { auth, db } from './firebase';

export class CalendarEventService {
  /**
   * Create calendar events for a goal
   * @param goalId Goal ID
   * @param events Array of calendar events to create
   */
  static async createCalendarEvents(goalId: string, events: Omit<CalendarEvent, 'id' | 'createdAt' | 'updatedAt'>[]): Promise<void> {
    if (!goalId || !events || events.length === 0) {
      throw new Error('Invalid parameters: goalId and events array are required');
    }

    const uid = auth.currentUser?.uid;
    if (!uid) throw new Error('Not authenticated');

    try {
      const batch = writeBatch(db);
      
      events.forEach(event => {
        const eventRef = doc(collection(db, 'users', uid, 'calendarEvents'));
        batch.set(eventRef, {
          ...event,
          createdAt: serverTimestamp(),
          updatedAt: serverTimestamp()
        });
      });
      
      await batch.commit();
      console.log(`[CalendarEventService] Created ${events.length} calendar events for user ${uid}`);
    } catch (error) {
      console.error('[CalendarEventService] Error creating calendar events:', error);
      throw new Error(`Failed to create calendar events: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Get calendar events for a goal within a date range
   * @param goalId Goal ID
   * @param startDate Start date (YYYY-MM-DD)
   * @param endDate End date (YYYY-MM-DD)
   */
  static async getCalendarEvents(goalId: string, startDate?: string, endDate?: string): Promise<CalendarEvent[]> {
    if (!goalId) {
      throw new Error('goalId is required');
    }

    const uid = auth.currentUser?.uid;
    if (!uid) throw new Error('Not authenticated');

    try {
      let q = query(
        collection(db, 'users', uid, 'calendarEvents'),
        where('goalId', '==', goalId),
        orderBy('date', 'asc')
      );

      if (startDate) {
        q = query(q, where('date', '>=', startDate));
      }
      if (endDate) {
        q = query(q, where('date', '<=', endDate));
      }

      const querySnapshot = await getDocs(q);
      const events: CalendarEvent[] = [];
      
      querySnapshot.forEach((doc) => {
        const data = doc.data();
        events.push({
          id: doc.id,
          date: data.date,
          time: data.time,
          goalId: data.goalId,
          source: data.source as 'single' | 'pattern',
          groupId: data.groupId,
          createdAt: data.createdAt?.toDate(),
          updatedAt: data.updatedAt?.toDate()
        });
      });

      console.log(`[CalendarEventService] Retrieved ${events.length} calendar events for user ${uid}`);
      return events;
    } catch (error) {
      console.error('[CalendarEventService] Error getting calendar events:', error);
      throw new Error(`Failed to get calendar events: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Update calendar events for a goal (preserves source field)
   * @param goalId Goal ID
   * @param events Array of calendar events to update
   */
  static async updateCalendarEvents(goalId: string, events: CalendarEvent[]): Promise<void> {
    if (!goalId || !events || events.length === 0) {
      throw new Error('Invalid parameters: goalId and events array are required');
    }

    const uid = auth.currentUser?.uid;
    if (!uid) throw new Error('Not authenticated');

    try {
      const batch = writeBatch(db);
      
      events.forEach(event => {
        if (!event.id) {
          throw new Error('Event ID is required for updates');
        }
        const eventRef = doc(db, 'users', uid, 'calendarEvents', event.id);
        batch.update(eventRef, {
          date: event.date,
          time: event.time,
          source: event.source, // Preserve source field
          groupId: event.groupId, // Preserve groupId field
          updatedAt: serverTimestamp()
        });
      });
      
      await batch.commit();
      console.log(`[CalendarEventService] Updated ${events.length} calendar events for user ${uid}`);
    } catch (error) {
      console.error('[CalendarEventService] Error updating calendar events:', error);
      throw new Error(`Failed to update calendar events: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Delete calendar events for a goal
   * @param goalId Goal ID
   * @param eventIds Array of event IDs to delete
   */
  static async deleteCalendarEvents(goalId: string, eventIds: string[]): Promise<void> {
    if (!goalId || !eventIds || eventIds.length === 0) {
      throw new Error('Invalid parameters: goalId and eventIds array are required');
    }

    const uid = auth.currentUser?.uid;
    if (!uid) throw new Error('Not authenticated');

    try {
      const batch = writeBatch(db);
      
      eventIds.forEach(eventId => {
        if (!eventId) {
          throw new Error('Event ID is required for deletion');
        }
        const eventRef = doc(db, 'users', uid, 'calendarEvents', eventId);
        batch.delete(eventRef);
      });
      
      await batch.commit();
      console.log(`[CalendarEventService] Deleted ${eventIds.length} calendar events for user ${uid}`);
    } catch (error) {
      console.error('[CalendarEventService] Error deleting calendar events:', error);
      throw new Error(`Failed to delete calendar events: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Create pattern events for a goal (replaces weekly schedule)
   * @param goalId Goal ID
   * @param options Pattern creation options
   */
  static async createPatternEvents(goalId: string, options: {
    startDate: string; 
    endDate: string;
    weekdays: number[];  // 0~6
    time: string;        // 'HH:mm'
  }): Promise<string> {
    const groupId = uuidv4();
    const { startDate, endDate, weekdays, time } = options;

    const start = new Date(startDate);
    const end = new Date(endDate);
    const ymd = (d: Date) => d.toISOString().slice(0,10);

    const events: Omit<CalendarEvent, 'id' | 'createdAt' | 'updatedAt'>[] = [];
    for (let d = new Date(start); d <= end; d.setDate(d.getDate() + 1)) {
      if (weekdays.includes(d.getDay())) {
        events.push({
          goalId,
          date: ymd(d),
          time,
          source: 'pattern',
          groupId,
        });
      }
    }
    if (events.length) await this.createCalendarEvents(goalId, events);
    return groupId;
  }

  /**
   * Delete pattern group events
   * @param goalId Goal ID
   * @param groupId Pattern group ID
   */
  static async deletePatternGroup(goalId: string, groupId: string): Promise<void> {
    const all = await this.getCalendarEvents(goalId);
    const ids = all.filter(e => e.groupId === groupId).map(e => e.id).filter(Boolean) as string[];
    if (ids.length) await this.deleteCalendarEvents(goalId, ids);
  }

  /**
   * Get calendar events grouped by source
   * @param goalId Goal ID
   * @param startDate Start date (YYYY-MM-DD)
   * @param endDate End date (YYYY-MM-DD)
   */
  static async getCalendarEventsBySource(
    goalId: string, 
    startDate?: string, 
    endDate?: string
  ): Promise<{ single: CalendarEvent[], pattern: CalendarEvent[] }> {
    const events = await CalendarEventService.getCalendarEvents(goalId, startDate, endDate);
    
    return {
      single: events.filter(e => e.source === 'single'),
      pattern: events.filter(e => e.source === 'pattern')
    };
  }
}
