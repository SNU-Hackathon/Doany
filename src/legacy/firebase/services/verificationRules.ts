import { VERIFICATION_DEFAULTS } from '../config/verification';
import { type GoalType } from '../constants/verificationPolicy';
import type { VerificationSignals } from '../types/firestore';
import { isWithin } from '../utils/time';

function withinWindow(t?: { present?: boolean; windowStart?: number | null; windowEnd?: number | null }, toleranceMinutes: number = VERIFICATION_DEFAULTS.timeToleranceMinutes) {
  if (!t?.present) return false;
  if (!t.windowStart || !t.windowEnd) return true;
  
  const now = Date.now();
  const toleranceMs = toleranceMinutes * 60 * 1000;
  const effectiveStart = t.windowStart - toleranceMs;
  const effectiveEnd = t.windowEnd + toleranceMs;
  
  return isWithin(now, effectiveStart, effectiveEnd);
}

// Schedule: Time AND (Manual+Location OR Time+Photo OR Time+Location OR Time+Manual)
export function evalScheduleRule(sig: VerificationSignals) {
  const timeOk = withinWindow(sig.time);
  const manualLocOk = !!(sig.manual?.present && sig.location?.present && sig.location.inside);
  const photoOk = !!(sig.photo?.present && 
    sig.photo.validationResult?.timeValid && 
    sig.photo.validationResult?.freshnessValid);
  const timeLocOk = !!(sig.time?.present && sig.location?.present && sig.location.inside);
  const timeManualOk = !!(sig.time?.present && sig.manual?.present);
  
  const either = manualLocOk || photoOk || timeLocOk || timeManualOk;
  return { 
    pass: timeOk && either, 
    details: { 
      timeOk, 
      manualLocOk, 
      photoOk,
      timeLocOk,
      timeManualOk,
      photoTimeValid: sig.photo?.validationResult?.timeValid,
      photoFreshValid: sig.photo?.validationResult?.freshnessValid
    } 
  };
}

// Frequency: (Manual+Location OR Manual+Photo)
export function evalFrequencyRule(sig: VerificationSignals) {
  const manualPhotoOk = !!(sig.manual?.present && 
    sig.photo?.present && 
    sig.photo.validationResult?.freshnessValid);
  const manualLocOk = !!(sig.manual?.present && sig.location?.present && sig.location.inside);
  
  return { 
    pass: manualLocOk || manualPhotoOk, 
    details: { 
      manualLocOk, 
      manualPhotoOk,
      photoFreshValid: sig.photo?.validationResult?.freshnessValid,
      photoLocationValid: sig.photo?.validationResult?.locationValid
    } 
  };
}

// Partner: Partner verification required
export function evalPartnerRule(sig: VerificationSignals) {
  const partnerOk = !!(sig.partner?.reviewed && sig.partner?.approved);
  const manualOk = !!sig.manual?.present;
  const photoOk = !!(sig.photo?.present && 
    sig.photo.validationResult?.timeValid && 
    sig.photo.validationResult?.freshnessValid);
  const locationOk = !!(sig.location?.present && sig.location.inside);
  
  // Partner is always required, optionally combine with others
  const hasAdditionalVerification = manualOk || photoOk || locationOk;
  
  return { 
    pass: partnerOk && (hasAdditionalVerification || true), // Partner alone is sufficient
    details: { 
      partnerOk,
      manualOk,
      photoOk,
      locationOk,
      hasAdditionalVerification
    } 
  };
}

export function evaluateByGoalType(goalType: GoalType, sig: VerificationSignals) {
  switch (goalType) {
    case 'schedule':
      return evalScheduleRule(sig);
    case 'frequency':
      return evalFrequencyRule(sig);
    case 'milestone':
      return evalScheduleRule(sig);
    default:
      return evalFrequencyRule(sig); // Default fallback
  }
}

// Inline smoke tests for rule validation
if (typeof process !== 'undefined' && process.env.NODE_ENV === 'development') {
  // Test Schedule rule: should FAIL with old photo
  const scheduleTestOldPhoto = evalScheduleRule({
    time: { present: true },
    photo: { 
      present: true,
      validationResult: { timeValid: false, freshnessValid: false } // old photo
    }
  });
  console.assert(!scheduleTestOldPhoto.pass, 'Schedule rule should FAIL with old photo');

  // Test Schedule rule: should PASS with fresh, time-valid photo
  const scheduleTestGoodPhoto = evalScheduleRule({
    time: { present: true },
    photo: { 
      present: true,
      validationResult: { timeValid: true, freshnessValid: true } // good photo
    }
  });
  console.assert(scheduleTestGoodPhoto.pass, 'Schedule rule should PASS with valid photo');

  // Test Frequency rule: should FAIL with manual+old photo
  const freqTestOldPhoto = evalFrequencyRule({
    manual: { present: true },
    photo: { 
      present: true,
      validationResult: { freshnessValid: false } // old photo
    }
  });
  console.assert(!freqTestOldPhoto.pass, 'Frequency rule should FAIL with old photo');

  console.log('[VerificationRules] âœ… Smoke tests passed');
}