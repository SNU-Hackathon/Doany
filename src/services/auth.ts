/**
 * Auth Service - Stub
 * 
 * Auth is handled by useAuth hook
 */

export class AuthError extends Error {
  code: string;
  friendlyMessage: string;
  suggestedAction?: string;

  constructor(message: string, code: string = 'AUTH_ERROR', friendlyMessage?: string, suggestedAction?: string) {
    super(message);
    this.code = code;
    this.friendlyMessage = friendlyMessage || message;
    this.suggestedAction = suggestedAction;
    this.name = 'AuthError';
  }
}

export async function sendReset(email: string): Promise<void> {
  console.warn('[sendReset] Not implemented yet');
  throw new Error('sendReset not implemented');
}

