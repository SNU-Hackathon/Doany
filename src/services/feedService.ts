/**
 * Feed Service - REST API Stub
 * 
 * TODO: Migrate to use REST API from src/api/feed.ts
 * This is a temporary stub to avoid compilation errors
 */

import { FeedPage, FeedPost } from '../types/feed';

export async function fetchFeedPage(options?: any): Promise<FeedPage> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Use getFeedGoals from src/api/feed.ts
  return {
    items: [],
    cursor: null,
    hasMore: false,
  };
}

export async function getFeedPost(postId: string): Promise<FeedPost | null> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Implement feed detail endpoint in REST API
  return null;
}

export async function getUserReaction(postId: string, userId: string): Promise<any> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Use feed endpoints from src/api/feed.ts
  return null;
}

export async function fetchComments(postId: string, cursor?: any): Promise<any> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Implement comment endpoints in REST API
  return { items: [], cursor: null, hasMore: false };
}

export async function addComment(postId: string, text: string, userId: string): Promise<void> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Implement comment endpoints in REST API
  throw new Error('addComment - Not yet migrated to REST API');
}

export async function deleteComment(commentId: string): Promise<void> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Implement comment endpoints in REST API
  throw new Error('deleteComment - Not yet migrated to REST API');
}

export async function createFeedPost(postData: any): Promise<string> {
  console.warn('[feedService] Using legacy stub - migrate to REST API');
  // TODO: Implement feed post creation in REST API
  throw new Error('createFeedPost - Not yet migrated to REST API');
}

