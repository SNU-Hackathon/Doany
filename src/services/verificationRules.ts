import type { GoalType, VerificationSignals } from '../types/firestore';

function withinWindow(t?: { now: number; windowStart?: number | null; windowEnd?: number | null }) {
  if (!t) return true;
  const { now, windowStart, windowEnd } = t;
  if (windowStart != null && now < windowStart) return false;
  if (windowEnd != null && now > windowEnd) return false;
  return true;
}

// Schedule: Time AND (Manual+Location OR Time+Photo)
export function evalScheduleRule(sig: VerificationSignals) {
  const timeOk = withinWindow(sig.time);
  const manual = !!sig.manual?.present && sig.manual.pass !== false;
  const loc = !!sig.location?.inside;
  const photo = !!sig.photo?.present;
  const either = (manual && loc) || (timeOk && photo);
  return { pass: timeOk && either, details: { timeOk, manual, loc, photo } };
}

// Frequency: (Manual+Location OR Manual+Photo)
export function evalFrequencyRule(sig: VerificationSignals) {
  const manual = !!sig.manual?.present && sig.manual.pass !== false;
  const loc = !!sig.location?.inside;
  const photo = !!sig.photo?.present;
  const either = (manual && loc) || (manual && photo);
  return { pass: either, details: { manual, loc, photo } };
}

export function evaluateByGoalType(goalType: GoalType, sig: VerificationSignals) {
  return goalType === 'schedule' ? evalScheduleRule(sig) : evalFrequencyRule(sig);
}