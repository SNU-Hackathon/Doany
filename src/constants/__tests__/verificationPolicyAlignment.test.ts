/**
 * Tests to ensure verification policy alignment between prompts and rules engine
 */

import { describe, expect, it } from 'vitest';
import { evalFrequencyRule, evalPartnerRule, evalScheduleRule } from '../../services/verificationRules';
import {
    ALLOWED_SIGNAL_COMBINATIONS,
    VERIFICATION_POLICIES,
    getExamplesForPrompt,
    getPolicyDescriptionForPrompt,
    isAllowedSignalCombination,
    validateVerificationSignals
} from '../verificationPolicy';

describe('Verification Policy Alignment', () => {
  describe('Policy Constants', () => {
    it('should have consistent signal combinations for schedule goals', () => {
      expect(VERIFICATION_POLICIES.schedule.withTimeAndPlace).toEqual(['time', 'location']);
      expect(VERIFICATION_POLICIES.schedule.withTimeOnly).toEqual(['time', 'photo']);
      expect(VERIFICATION_POLICIES.schedule.fallback).toEqual(['time', 'manual']);
    });

    it('should have consistent signal combinations for frequency goals', () => {
      expect(VERIFICATION_POLICIES.frequency.primary).toEqual(['manual', 'photo']);
      expect(VERIFICATION_POLICIES.frequency.withLocation).toEqual(['manual', 'location']);
      expect(VERIFICATION_POLICIES.frequency.fallback).toEqual(['manual']);
    });

    it('should have consistent signal combinations for partner goals', () => {
      expect(VERIFICATION_POLICIES.partner.required).toEqual(['partner']);
      expect(VERIFICATION_POLICIES.partner.withOthers).toEqual(['partner', 'manual']);
    });
  });

  describe('Prompt Integration', () => {
    it('should generate policy description for prompts', () => {
      const description = getPolicyDescriptionForPrompt();
      
      expect(description).toContain('VERIFICATION SIGNALS POLICY');
      expect(description).toContain('Schedule with time+place');
      expect(description).toContain('Schedule with time only');
      expect(description).toContain('Frequency goals');
      expect(description).toContain('Partner type');
      
      // Should include actual policy values
      expect(description).toContain(JSON.stringify(VERIFICATION_POLICIES.schedule.withTimeAndPlace));
      expect(description).toContain(JSON.stringify(VERIFICATION_POLICIES.schedule.withTimeOnly));
      expect(description).toContain(JSON.stringify(VERIFICATION_POLICIES.frequency.primary));
      expect(description).toContain(JSON.stringify(VERIFICATION_POLICIES.partner.required));
    });

    it('should generate examples for prompts', () => {
      const examples = getExamplesForPrompt();
      
      expect(examples).toContain('EXAMPLES');
      expect(examples).toContain('GOOD:');
      
      // Should include actual policy values in examples
      expect(examples).toContain(JSON.stringify(VERIFICATION_POLICIES.schedule.withTimeOnly));
      expect(examples).toContain(JSON.stringify(VERIFICATION_POLICIES.frequency.primary));
      expect(examples).toContain(JSON.stringify(VERIFICATION_POLICIES.partner.required));
    });
  });

  describe('Policy Validation', () => {
    it('should validate schedule verification signals correctly', () => {
      // Valid combinations
      expect(validateVerificationSignals('schedule', ['time', 'location']).valid).toBe(true);
      expect(validateVerificationSignals('schedule', ['time', 'photo']).valid).toBe(true);
      expect(validateVerificationSignals('schedule', ['time', 'manual']).valid).toBe(true);
      
      // Invalid combinations
      const invalidResult = validateVerificationSignals('schedule', ['location', 'photo']);
      expect(invalidResult.valid).toBe(false);
      expect(invalidResult.errors).toContain('Schedule goals must include "time" signal');
    });

    it('should validate frequency verification signals correctly', () => {
      // Valid combinations
      expect(validateVerificationSignals('frequency', ['manual', 'photo']).valid).toBe(true);
      expect(validateVerificationSignals('frequency', ['manual', 'location']).valid).toBe(true);
      
      // Invalid combinations
      const invalidResult = validateVerificationSignals('frequency', ['photo', 'location']);
      expect(invalidResult.valid).toBe(false);
      expect(invalidResult.errors).toContain('Frequency goals must include "manual" signal');
    });

    it('should validate partner verification signals correctly', () => {
      // Valid combinations
      expect(validateVerificationSignals('partner', ['partner']).valid).toBe(true);
      expect(validateVerificationSignals('partner', ['partner', 'manual']).valid).toBe(true);
      
      // Invalid combinations
      const invalidResult = validateVerificationSignals('partner', ['manual', 'photo']);
      expect(invalidResult.valid).toBe(false);
      expect(invalidResult.errors).toContain('Partner goals must include "partner" signal');
    });
  });

  describe('Allowed Signal Combinations', () => {
    it('should have comprehensive allowed combinations', () => {
      // Schedule combinations
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['time', 'location']);
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['time', 'photo']);
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['time', 'manual']);
      
      // Frequency combinations
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['manual', 'photo']);
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['manual', 'location']);
      
      // Partner combinations
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['partner']);
      expect(ALLOWED_SIGNAL_COMBINATIONS).toContainEqual(['partner', 'manual']);
    });

    it('should validate allowed combinations correctly', () => {
      // Valid combinations
      expect(isAllowedSignalCombination(['time', 'location'])).toBe(true);
      expect(isAllowedSignalCombination(['manual', 'photo'])).toBe(true);
      expect(isAllowedSignalCombination(['partner'])).toBe(true);
      
      // Invalid combinations
      expect(isAllowedSignalCombination(['location', 'photo'])).toBe(false);
      expect(isAllowedSignalCombination(['time'])).toBe(false);
      expect(isAllowedSignalCombination(['manual'])).toBe(true); // Fallback is allowed
    });
  });

  describe('Rules Engine Alignment', () => {
    // Mock verification signals for testing
    const createMockSignals = (signals: string[]) => ({
      time: signals.includes('time') ? { present: true, windowStart: Date.now() - 1000, windowEnd: Date.now() + 1000 } : undefined,
      location: signals.includes('location') ? { present: true, inside: true } : undefined,
      photo: signals.includes('photo') ? { present: true, validationResult: { timeValid: true, freshnessValid: true } } : undefined,
      manual: signals.includes('manual') ? { present: true } : undefined,
      partner: signals.includes('partner') ? { reviewed: true, approved: true } : undefined,
    });

    it('should align schedule rules with policy', () => {
      // Test policy-compliant combinations
      const timeLocationSignals = createMockSignals(['time', 'location']);
      const timeLocationResult = evalScheduleRule(timeLocationSignals);
      expect(timeLocationResult.pass).toBe(true);
      
      const timePhotoSignals = createMockSignals(['time', 'photo']);
      const timePhotoResult = evalScheduleRule(timePhotoSignals);
      expect(timePhotoResult.pass).toBe(true);
      
      const timeManualSignals = createMockSignals(['time', 'manual']);
      const timeManualResult = evalScheduleRule(timeManualSignals);
      expect(timeManualResult.pass).toBe(true);
    });

    it('should align frequency rules with policy', () => {
      // Test policy-compliant combinations
      const manualPhotoSignals = createMockSignals(['manual', 'photo']);
      const manualPhotoResult = evalFrequencyRule(manualPhotoSignals);
      expect(manualPhotoResult.pass).toBe(true);
      
      const manualLocationSignals = createMockSignals(['manual', 'location']);
      const manualLocationResult = evalFrequencyRule(manualLocationSignals);
      expect(manualLocationResult.pass).toBe(true);
    });

    it('should align partner rules with policy', () => {
      // Test policy-compliant combinations
      const partnerSignals = createMockSignals(['partner']);
      const partnerResult = evalPartnerRule(partnerSignals);
      expect(partnerResult.pass).toBe(true);
      
      const partnerManualSignals = createMockSignals(['partner', 'manual']);
      const partnerManualResult = evalPartnerRule(partnerManualSignals);
      expect(partnerManualResult.pass).toBe(true);
    });

    it('should reject policy-violating combinations', () => {
      // Schedule without time should fail
      const locationOnlySignals = createMockSignals(['location']);
      const locationOnlyResult = evalScheduleRule(locationOnlySignals);
      expect(locationOnlyResult.pass).toBe(false);
      
      // Frequency without manual should fail
      const photoOnlySignals = createMockSignals(['photo']);
      const photoOnlyResult = evalFrequencyRule(photoOnlySignals);
      expect(photoOnlyResult.pass).toBe(false);
      
      // Partner without partner signal should fail
      const manualOnlySignals = createMockSignals(['manual']);
      const manualOnlyResult = evalPartnerRule(manualOnlySignals);
      expect(manualOnlyResult.pass).toBe(false);
    });
  });

  describe('Policy Snapshot Tests', () => {
    it('should maintain consistent policy structure', () => {
      // Snapshot test for policy structure
      const policySnapshot = {
        schedule: {
          withTimeAndPlace: VERIFICATION_POLICIES.schedule.withTimeAndPlace,
          withTimeOnly: VERIFICATION_POLICIES.schedule.withTimeOnly,
          fallback: VERIFICATION_POLICIES.schedule.fallback,
        },
        frequency: {
          primary: VERIFICATION_POLICIES.frequency.primary,
          withLocation: VERIFICATION_POLICIES.frequency.withLocation,
          fallback: VERIFICATION_POLICIES.frequency.fallback,
        },
        partner: {
          required: VERIFICATION_POLICIES.partner.required,
          withOthers: VERIFICATION_POLICIES.partner.withOthers,
        },
      };
      
      // This test will fail if the policy structure changes
      expect(policySnapshot).toMatchSnapshot();
    });

    it('should maintain consistent allowed combinations', () => {
      // Snapshot test for allowed combinations
      expect(ALLOWED_SIGNAL_COMBINATIONS).toMatchSnapshot();
    });

    it('should maintain consistent policy descriptions', () => {
      // Snapshot test for policy descriptions
      const descriptionSnapshot = {
        promptDescription: getPolicyDescriptionForPrompt(),
        examples: getExamplesForPrompt(),
      };
      
      expect(descriptionSnapshot).toMatchSnapshot();
    });
  });

  describe('Cross-Validation', () => {
    it('should ensure all policy combinations are in allowed combinations', () => {
      const allPolicyCombinations = [
        ...VERIFICATION_POLICIES.schedule.withTimeAndPlace,
        ...VERIFICATION_POLICIES.schedule.withTimeOnly,
        ...VERIFICATION_POLICIES.schedule.fallback,
        ...VERIFICATION_POLICIES.frequency.primary,
        ...VERIFICATION_POLICIES.frequency.withLocation,
        ...VERIFICATION_POLICIES.frequency.fallback,
        ...VERIFICATION_POLICIES.partner.required,
        ...VERIFICATION_POLICIES.partner.withOthers,
      ];
      
      // Each policy combination should be in allowed combinations
      const policyCombinationArrays = [
        VERIFICATION_POLICIES.schedule.withTimeAndPlace,
        VERIFICATION_POLICIES.schedule.withTimeOnly,
        VERIFICATION_POLICIES.schedule.fallback,
        VERIFICATION_POLICIES.frequency.primary,
        VERIFICATION_POLICIES.frequency.withLocation,
        VERIFICATION_POLICIES.frequency.fallback,
        VERIFICATION_POLICIES.partner.required,
        VERIFICATION_POLICIES.partner.withOthers,
      ];
      
      policyCombinationArrays.forEach(combination => {
        expect(isAllowedSignalCombination(combination)).toBe(true);
      });
    });

    it('should ensure validation passes for all policy combinations', () => {
      // Test schedule combinations
      expect(validateVerificationSignals('schedule', VERIFICATION_POLICIES.schedule.withTimeAndPlace).valid).toBe(true);
      expect(validateVerificationSignals('schedule', VERIFICATION_POLICIES.schedule.withTimeOnly).valid).toBe(true);
      expect(validateVerificationSignals('schedule', VERIFICATION_POLICIES.schedule.fallback).valid).toBe(true);
      
      // Test frequency combinations
      expect(validateVerificationSignals('frequency', VERIFICATION_POLICIES.frequency.primary).valid).toBe(true);
      expect(validateVerificationSignals('frequency', VERIFICATION_POLICIES.frequency.withLocation).valid).toBe(true);
      expect(validateVerificationSignals('frequency', VERIFICATION_POLICIES.frequency.fallback).valid).toBe(true);
      
      // Test partner combinations
      expect(validateVerificationSignals('partner', VERIFICATION_POLICIES.partner.required).valid).toBe(true);
      expect(validateVerificationSignals('partner', VERIFICATION_POLICIES.partner.withOthers).valid).toBe(true);
    });
  });
});
