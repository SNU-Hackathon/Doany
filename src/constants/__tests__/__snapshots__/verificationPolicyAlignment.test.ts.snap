// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Verification Policy Alignment > Policy Snapshot Tests > should maintain consistent allowed combinations 1`] = `
[
  [
    "time",
    "location",
  ],
  [
    "time",
    "photo",
  ],
  [
    "time",
    "manual",
  ],
  [
    "time",
    "location",
    "photo",
  ],
  [
    "time",
    "location",
    "manual",
  ],
  [
    "time",
    "photo",
    "manual",
  ],
  [
    "manual",
    "photo",
  ],
  [
    "manual",
    "location",
  ],
  [
    "manual",
    "photo",
    "location",
  ],
  [
    "partner",
  ],
  [
    "partner",
    "manual",
  ],
  [
    "partner",
    "photo",
  ],
  [
    "partner",
    "location",
  ],
  [
    "partner",
    "manual",
    "photo",
  ],
  [
    "partner",
    "manual",
    "location",
  ],
  [
    "partner",
    "photo",
    "location",
  ],
  [
    "manual",
  ],
]
`;

exports[`Verification Policy Alignment > Policy Snapshot Tests > should maintain consistent policy descriptions 1`] = `
{
  "examples": "EXAMPLES:
GOOD: {"type":"schedule","originalText":"월수금 6시 러닝","schedule":{"events":[{"dayOfWeek":"mon","time":"06:00"},{"dayOfWeek":"wed","time":"06:00"},{"dayOfWeek":"fri","time":"06:00"}]},"verification":{"signals":["time","photo"]}}
GOOD: {"type":"frequency","originalText":"일주일에 3번 독서","frequency":{"targetPerWeek":3,"windowDays":7},"verification":{"signals":["manual","photo"]}}
GOOD: {"type":"partner","originalText":"매일 코치와 운동 검토","partner":{"required":true,"name":"코치"},"verification":{"signals":["partner"]}}",
  "promptDescription": "VERIFICATION SIGNALS POLICY:
- Schedule with time+place: ["time","location"]
- Schedule with time only: ["time","photo"] or ["time","manual"]
- Frequency goals: ["manual","photo"] (add location when meaningful)
- Partner type: MUST include ["partner"], optionally combine with others if independently verifiable",
}
`;

exports[`Verification Policy Alignment > Policy Snapshot Tests > should maintain consistent policy structure 1`] = `
{
  "frequency": {
    "fallback": [
      "manual",
    ],
    "primary": [
      "manual",
      "photo",
    ],
    "withLocation": [
      "manual",
      "location",
    ],
  },
  "partner": {
    "required": [
      "partner",
    ],
    "withOthers": [
      "partner",
      "manual",
    ],
  },
  "schedule": {
    "fallback": [
      "time",
      "manual",
    ],
    "withTimeAndPlace": [
      "time",
      "location",
    ],
    "withTimeOnly": [
      "time",
      "photo",
    ],
  },
}
`;
